src/Features/CSharp/Portable/ConvertLinq/ConvertForEachToLinqQuery/CSharpConvertForEachToLinqQueryProvider.cs

```diff
@@ -0,0 +1,340 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Composition;
+using System.Linq;
+using System.Threading;
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.ConvertLinq.ConvertForEachToLinqQuery;
+using Microsoft.CodeAnalysis.CSharp.Extensions;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Roslyn.Utilities;
+using SyntaxNodeOrTokenExtensions = Microsoft.CodeAnalysis.Shared.Extensions.SyntaxNodeOrTokenExtensions;
+
+namespace Microsoft.CodeAnalysis.CSharp.ConvertLinq.ConvertForEachToLinqQuery
+{
+    [ExportCodeRefactoringProvider(LanguageNames.CSharp, Name = nameof(CSharpConvertForEachToLinqQueryProvider)), Shared]
+    internal sealed class CSharpConvertForEachToLinqQueryProvider
+        : AbstractConvertForEachToLinqQueryProvider<ForEachStatementSyntax, StatementSyntax>
+    {
+        protected override IConverter<ForEachStatementSyntax, StatementSyntax> CreateDefaultConverter(
+            ForEachInfo<ForEachStatementSyntax, StatementSyntax> forEachInfo)
+            => new DefaultConverter(forEachInfo);
+
+        protected override ForEachInfo<ForEachStatementSyntax, StatementSyntax> CreateForEachInfo(
+            ForEachStatementSyntax forEachStatement,
+            SemanticModel semanticModel)
+        {
+            var identifiersBuilder = ImmutableArray.CreateBuilder<SyntaxToken>();
+            identifiersBuilder.Add(forEachStatement.Identifier);
+            var convertingNodesBuilder = ImmutableArray.CreateBuilder<ExtendedSyntaxNode>();
+            IEnumerable<StatementSyntax> statementsCannotBeConverted = null;
```

Comment:
>statementsCannotBeConverted [](start = 41, length = 27)

Can this be replaced with an index such as `indexOfFirstStatementThatCannotBeConverted`?

Reply:
Never mind, figured out this can be a nested statement.

---
In reply to: [196947102](https://github.com/dotnet/roslyn/pull/27130#discussion_r196947102) [](ancestors = 196947102)